File: gatelogue-types-ts/src/rail.ts
import { Node } from "./node.js";
import { LocatedNode } from "./located.js";

export type RailMode = "warp" | "cart" | "traincarts" | "vehicles";

export class RailCompany extends Node {
  get name(): string {
    return this.getColumn("RailCompany", "name");
  }
  get link(): string | null {
    return this.getColumn("RailCompany", "link");
  }

  get lines(): RailLine[] {
    return this.getDerived(RailLine, "rail/company_lines");
  }
  get stations(): RailStation[] {
    return this.getDerived(RailStation, "rail/company_stations");
  }
  get platforms(): RailPlatform[] {
    return this.getDerived(RailPlatform, "rail/company_platforms");
  }
}

export class RailLine extends Node {
  get code(): string {
    return this.getColumn("RailLine", "code");
  }
  get company(): RailCompany {
    return this.getColumnFK("RailLine", "company", RailCompany)!;
  }
  get name(): string | null {
    return this.getColumn("RailLine", "name");
  }
  get colour(): string | null {
    return this.getColumn("RailLine", "colour");
  }
  get mode(): RailMode | null {
    return this.getColumn("RailLine", "name");
  }
  get local(): boolean | null {
    const result = this.getColumn<number>("RailLine", "name");
    return result === null ? null : result === 1;
  }

  get platforms(): RailPlatform[] {
    return this.getDerived(RailPlatform, "rail/line_platforms");
  }
  get stations(): RailStation[] {
    return this.getDerived(RailStation, "rail/line_stations");
  }
}

export class RailStation extends LocatedNode {
  get codes(): string[] {
    return this.getSet("RailStationCodes", "code");
  }
  get company(): RailCompany {
    return this.getColumnFK("RailStation", "company", RailCompany)!;
  }
  get name(): string | null {
    return this.getColumn("RailStation", "name");
  }

  get platforms(): RailPlatform[] {
    return this.getDerived(RailPlatform, "rail/station_platforms");
  }
  get connectionsFromHere(): RailConnection[] {
    return this.getDerived(
      RailConnection,
      "rail/station_connections_from_here",
    );
  }
  get connectionsToHere(): RailConnection[] {
    return this.getDerived(RailConnection, "rail/station_connections_to_here");
  }
  get lines(): RailLine[] {
    return this.getDerived(RailLine, "rail/station_lines");
  }
}

export class RailPlatform extends Node {
  get code(): string | null {
    return this.getColumn("RailPlatform", "code");
  }
  get station(): RailStation {
    return this.getColumnFK("RailPlatform", "station", RailStation)!;
  }

  get connectionsFromHere(): RailConnection[] {
    return this.getDerived(
      RailConnection,
      "rail/platform_connections_from_here",
    );
  }
  get connectionsToHere(): RailConnection[] {
    return this.getDerived(RailConnection, "rail/platform_connections_to_here");
  }
  get lines(): RailLine[] {
    return this.getDerived(RailLine, "rail/platform_lines");
  }
}

export class RailConnection extends Node {
  get line(): RailLine {
    return this.getColumnFK("RailConnection", "line", RailLine)!;
  }
  get from(): RailPlatform {
    return this.getColumnFK("RailConnection", "from", RailPlatform)!;
  }
  get to(): RailPlatform {
    return this.getColumnFK("RailConnection", "to", RailPlatform)!;
  }
  get direction(): string | null {
    return this.getColumn("RailConnection", "direction");
  }
  get duration(): number | null {
    return this.getColumn("RailConnection", "duration");
  }
}

File: gatelogue-types-rs/src/node/rail.rs
use strum_macros::EnumString;

use crate::{_from_sql_for_enum, _get_column, _get_derived_vec, _get_set, node_type, util::ID};

#[derive(Clone, Copy, PartialEq, Eq, Debug, EnumString)]
pub enum RailMode {
    #[strum(serialize = "warp")]
    Warp,
    #[strum(serialize = "cart")]
    Cart,
    #[strum(serialize = "traincarts")]
    TrainCarts,
    #[strum(serialize = "vehicles")]
    Vehicles,
}
_from_sql_for_enum!(RailMode);

node_type!(RailCompany);
impl RailCompany {
    _get_column!("RailCompany", name, String);
    _get_column!("RailCompany", link, Option<String>);
    _get_derived_vec!(lines, RailLine, "../sql/rail/company_lines.sql");
    _get_derived_vec!(stations, RailStation, "../sql/rail/company_stations.sql");
    _get_derived_vec!(platforms, RailPlatform, "../sql/rail/company_platforms.sql");
}

node_type!(RailLine);
impl RailLine {
    _get_column!("RailLine", code, String);
    _get_column!("RailLine", company, RailCompany);
    _get_column!("RailLine", name, Option<String>);
    _get_column!("RailLine", colour, Option<String>);
    _get_column!("RailLine", mode, Option<RailMode>);
    _get_column!("RailLine", local, Option<bool>);

    _get_derived_vec!(platforms, RailPlatform, "../sql/rail/line_platforms.sql");
    _get_derived_vec!(stations, RailStation, "../sql/rail/company_stations.sql");
}

node_type!(located RailStation);
impl RailStation {
    _get_set!("RailStationCodes", codes, "code", String);
    _get_column!("RailStation", company, RailCompany);
    _get_column!("RailStation", name, Option<String>);

    _get_derived_vec!(platforms, RailPlatform, "../sql/rail/station_platforms.sql");
    _get_derived_vec!(
        connections_from_here,
        RailConnection,
        "../sql/rail/station_connections_from_here.sql"
    );
    _get_derived_vec!(
        connections_to_here,
        RailConnection,
        "../sql/rail/station_connections_to_here.sql"
    );
    _get_derived_vec!(lines, RailLine, "../sql/rail/station_lines.sql");
}

node_type!(RailPlatform);
impl RailPlatform {
    _get_column!("RailPlatform", code, Option<String>);
    _get_column!("RailPlatform", station, RailStation);

    _get_derived_vec!(
        connections_from_here,
        RailConnection,
        "../sql/rail/platform_connections_from_here.sql"
    );
    _get_derived_vec!(
        connections_to_here,
        RailConnection,
        "../sql/rail/platform_connections_to_here.sql"
    );
    _get_derived_vec!(lines, RailLine, "../sql/rail/platform_lines.sql");
}

node_type!(RailConnection);
impl RailConnection {
    _get_column!("RailConnection", line, RailLine);
    _get_column!("RailConnection", from, RailPlatform);
    _get_column!("RailConnection", to, RailPlatform);
    _get_column!("RailConnection", direction, Option<String>);
    _get_column!("RailConnection", duration, Option<u32>);
}
